python/cudf/cudf/io/json.py

```diff
@@ -1,35 +1,101 @@
 # Copyright (c) 2019-2024, NVIDIA CORPORATION.
+from __future__ import annotations
 
+import os
 import warnings
 from collections import abc
 from io import BytesIO, StringIO
+from typing import TYPE_CHECKING, Any, Literal
 
 import numpy as np
 import pandas as pd
 
+import pylibcudf as plc
+
 import cudf
-from cudf._lib import json as libjson
+from cudf._lib.column import Column
+from cudf._lib.types import dtype_to_pylibcudf_type
+from cudf._lib.utils import _data_from_columns, data_from_pylibcudf_io
+from cudf.core.buffer import acquire_spill_lock
 from cudf.utils import ioutils
 from cudf.utils.dtypes import _maybe_convert_to_default_type
 
+if TYPE_CHECKING:
+    from cudf.core.column import ColumnBase
+
+
+def _add_df_col_struct_names(
+    df: cudf.DataFrame, child_names_dict: dict
+) -> None:
+    for name, child_names in child_names_dict.items():
+        col = df._data[name]
+        df._data[name] = _update_col_struct_field_names(col, child_names)
+
+
+def _update_col_struct_field_names(
+    col: ColumnBase, child_names: dict
+) -> ColumnBase:
+    if col.children:
+        children = list(col.children)
+        for i, (child, names) in enumerate(
+            zip(children, child_names.values())
+        ):
+            children[i] = _update_col_struct_field_names(child, names)
+        col.set_base_children(tuple(children))
+
+    if isinstance(col.dtype, cudf.StructDtype):
+        col = col._rename_fields(child_names.keys())  # type: ignore[attr-defined]
+
+    return col
```

Comment:
I guess it is OK that we update names in-place, but it is a minor potential footgun for two columns that share data (but in theory could have different names).